# Variables

### Clicker #0

Please answer this iClicker question. 

- a) okay, sure
- b) yes, of course!
- c) I don't want to
- d) why are we doing this
- e) I hate iClickers

## Programming With Python

Programming: a way to ask computer to store values (variables), and do things with them (operations).

### Variables and Comments

In [15]:
# This is a comment. You can write a comment by using a `#`
# A comment is a note for someone reading your code
# It doesn't get executed by the kernel

one_dozen = 12  # on the left of '=' is the variable name, on the right of '=' is the value stored in the variable.  To the right of # is a comment

bakers_dozen = 13 # Comments can be 'inline', like this one, or by themselves like the one at the top of this cell

darth_vader = "Luke's Father" # variables can be called anything you want, comments help people understand your code

## Declaring Variables

<div class="alert alert-success">
In programming, variables are things that store values. Variables are declared with `name = value`.
</div>

## The rules of declaring variables

- Names are always on the left of the `=`, values are always on the right
- Names are case sensitive
- Variables must start with letters
    - After that, they can include numbers, and underscores
    - They cannot include special characters (like &, *, #, etc)
- Python doesn't care what you name you variables
    - Humans do care. Pick names that describe the data or value that they store.
    - Make it so you don't have to explain your code to anyone.
        - They should just get it from reading the code.

## Spelling and grammar count.  

- Most errors when we're starting out are because of typos and indents. 
- Programming languages are unforgiving.
    - They don't care how close you are
    - They don't give you the benefit of the doubt
    - Right or wrong, they follow the instructions you type exactly
    - It is your responsibility as a coder to make sure your code is free from typos.


Every experiment we do in Psychology will have lots of variables.  Let's start with our mock participant in our mock experiment. After providing informed consent to an experiment participants are assigned an ID number to accompany their data instead of their names to provide anonymity.  In the example below `participant` is a variable that will contain their ID number. Remember that in jupyter notebooks, when you type the name of a variable, it displays the value. 

In [16]:
participant = 1331
participant

1331

Notice that I used lower-cased letters for `participant`.

- In Python, you should always use lower-cased letters for the names of youre variables. 
- This helps other people who read your code quickly identify the type of object they are looking at.
- If you need to make a variable that is more than one word, use an underscore `_` instead of a space.
- If you want to learn the rules of formatting your code, google "Python PEP-8"


- You can assign a variable to another variable

In [17]:
participant_number = participant
participant_number

1331

You can change the value stored in variables

In [18]:
participant += 1  # this means participant = participant + 1
participant

1332

This will also change the variable we assigned it to.  That's because when we wrote `participant_number = participant`, we didn't copy the value stored in `participant`, i.e. `1331`, instead, we pointed at the variable, so when `participant` changes, so does `participant_number`.

In [19]:
participant_number

1331

But when we change participant_number, it doesn't affect participant.  It's not a 2-way street.

In [20]:
participant_number += 1
print('participant_number: ', participant_number)
print('participant: ', participant)

participant_number:  1332
participant:  1332


### Clicker #1

What do you think the value of `participant_number` will be now if we add 1 to `participant` by executing the following code: 
`participant += 1`?

- a) 1331
- b) 1332
- c) 1333
- d) I'm totally lost, what's going on?
- e) I hate iClickers, please stop doing this.

In [21]:
participant += 1
print('participant_number: ', participant_number)
print('participant: ', participant)

participant_number:  1332
participant:  1333


So now `participant_number` and `participant` store different values. That can make things really weird unless you are doing this on purpose. Practical uses of this are keeping track of multiple items... but be careful and make sure you code this with intent. It's not really for beginners.

Your variables don't have to be about participants.  They can and should be about whatever they represent.  For math and graphing, we often use the single letters `x` and `y`.

In [4]:
x = 1
y = 2
x + y

3

You can use numbers and letters, but variables have to start with a letter.

In [6]:
r2d2 = 'white and blue robot'
c3p0 = 'gold robot'
print(r2d2,'\n', c3p0)

white and blue robot 
 gold robot


# Print

- The print function outputs things to the screen, not the printer
- When we run code in Jupyter notebook, it only shows the final output, like in the previous slide
- When we want to see more, or want to write code for use outside of Jupyter Notebooks, use "print"

## You can print the values stored in variables

In [43]:
apple = "Honey Crisp"
pi = 3.14 

In [44]:
print(apple)

Honey Crisp


In [45]:
print(pi)

3.14


In [46]:
print (apple, pi)

Honey Crisp 3.14


## You can print text

In [14]:
print("To be, or not to be, that is the question:\n" 
      "Whether 'tis nobler in the mind to suffer\n"
      "The slings and arrows of outrageous fortune,\n"
      "Or to take Arms against a Sea of troubles,\n"
      "And by opposing end them")


To be, or not to be, that is the question:
Whether 'tis nobler in the mind to suffer
The slings and arrows of outrageous fortune,
Or to take Arms against a Sea of troubles,
And by opposing end them


<div class="alert alert-info">
You can see I typed "\n" to tell the computer to make a new line.\ is called an escape string.  Google 'python escape strings' to learn more escape strings.
    </div>

## You can print text and variables together

In [12]:
r_squared = 0.76

# This defines another variable
effect_size = 'large'

print("The r^2 value is: ", r_squared, "\n", "The effect size is ", effect_size)

The r^2 value is:  0.76 
 The effect size is  large


## You can print a lot of things a lot of ways
We'll learn more about string formatting and string manipulations later.

# <font color="red">Warning - "print" is different in Python 3.X and Python 2.7</font>



## A lot of web help for Python is in Python 2.7
- In Python 2.7, which we don't use in class, you print like this:
    - `print "lorum ipsum"`
- In Python 3.X, the version we use in class, you print like this:
    - `print("lorum ipsum")`
- To fix the print statements, just make sure you use parentheses like this `print("this way")`.

# Back to Variables....

## Code Variables != Math Variables

In mathematics: `=` refers to equality (as a statement of truth).

In coding: `=` refers to assignment. You can read the '=' as 'gets'.  

x = 2 + 5 can be read as "x gets the value that results when you add 2 plus 5".

In coding, you can use `==` to evaluate equality.

Math: What is x?

`102 = 10x + 2`

Code: What is x?

`x = x + 1`

### Clicker #2

After executing the following code, what will be the value of `my_var`?

In [22]:
my_var = 2

my_var = my_var + 1

print(my_var)

3


- a) 2
- b) 3
- c) "my_var + 1"
- d) This code will fail
- e) I'm totally lost.  What is going on here?


### Reminder

- In programming `=` means assignment
- Anything to the right of the `=` is evaluated before assignment

## Namespace

<div class="alert alert-success">
The namespace is the 'place' where all your currently defined code is declared - all the things you have stored in active memory. 
</div>

In [17]:
# You can list everything declated in the namespace with '%whos'
%whos

Variable            Type    Data/Info
-------------------------------------
my_other_variable   int     13
my_var              int     1
my_variable         int     12
other_var           str     variables are cool
x                   int     1
y                   int     2


## <center> Don't reuse & recycle variable names</center>
<img src="img/recycle.jpg">


----------------
- Technically you can re-use variables.  
- It's bad to re-use variable names because it's easy to get mixed up. 
- If you make a variable by accident, or have a type, or just don't want it anymore, just leave it alone.  
    - It's not going to hurt anyone if you just don't type it.

### iClicker Question #2

After executing the following code, what will be the value of `var_2`?

In [18]:
var_2 = var_1 = 1

print(var_2)

1


- a) 'var_1'
- b) 1
- c) 2
- d) This code will fail

### Reminders

- There can be more than one assignment in a single line
- Anything to the right of the `=` is evaluated before assignment
    - This process proceeds from right to left

# Variable Types

<div class="alert alert-success">
Every variable has a 'type', which refers to the kind of variable that it is, and how the computer stores that data.
</div>

You don't have to tell Python what type of variable something is.  It figures it out itself by the nature of the contents. In other programming languages, you do have to specify this when declaring variables, so we are happy we are using Python so we can bypass this step.

In [23]:
# Declare a variable
variable_name = 1

# You can always ask Python 'what type is this variable' using:
type(variable_name)

int

### Int

<div class="alert alert-success">
`Integers` store whole numbers.
</div>

In [2]:
my_integer = 1
another_integer = 321
a_negative_integer = -345
zero = 0

### Float

<div class="alert alert-success">
`Floats` store signed, decimal-point numbers.  </div>
The decimal point 'floats' around, depending on how many decimal places you need.
They don't have a signature, they are signed (i.e.) positive (+) or negative (-).


In [10]:
my_float = 1.0
another_float = -231.45

### String

<div class="alert alert-success">
`Strings` store characters, as text.  
</div>


In [16]:
my_string = 'words, words, words'
another_string = 'more words'

# Note that strings can be defined with either '' or ""
and_another = "and some more"

<div class="alert alert-success">
`Strings` can also store numbers, as text.  
</div>

In [24]:
one = "1"
two = "2"
one + two

'12'

## Boolean

### `Booleans` store `True` or `False`. 
When a variable is:
- `False`
- empty string (e.g. `''`)
- or `0`
    - Python recognizes it as False.


When a variable is:
- `True`
- empty string (e.g. `' '`)
- or not `0`
    - Python recognizes it as True.



## iClicker Question #3


When a variable is `'False'`

e.g. `my_var = 'False'`

Does Python recognize this variable as   
1) `True`   
2) `False`   
3) I'm lost, what's going on?  

In [41]:
# iClicker #3 solution
my_var = 'False'

In [42]:
my_var is True

False

In [43]:
my_var is False

False

In [44]:
my_var is 'False'

True

## None

<div class="alert alert-success">
`None` is a special type that stores `None`, used to denote a null or empty value.
</div>

In [47]:
the_concept_of_nothing = None

In [48]:
the_concept_of_nothing is True

False

In [49]:
the_concept_of_nothing is False

False

In [50]:
the_concept_of_nothing is None

True

That's deep.

### Reminders

- Multiple variables can relate to the same value(s)
    - e.g. `x = 3` & `y = 3`can both peacefully co-exist